src/data/scripts/hullmods/vayra_modular_shields.java

```diff
@@ -116,68 +122,200 @@ public void applyEffectsAfterShipCreation(ShipAPI ship, String id) {
-                    doShieldEmitterStuff(ship);
+                    // Handle emitters
+                    doShieldEmitterStuff(ship, generator);
                     break;
-                } // handle generator
+                }
                 case SHIELD_GENERATOR_ID: {
-                    doShieldGenStuff(ship);
+                    // handle generator
+                    doShieldGenStuff(ship, emitters, engine);
                     break;
-                } // handle damage reduction and visuals for everything else
+                }
                 default: {
-                    doEveryOtherPartStuff(ship, amount, engine);
+                    // handle damage reduction and visuals for everything else
+                    doEveryOtherPartStuff(ship, amount, generator, engine);
                     break;
                 }
             }
         }
     }
 
-    private void doEveryOtherPartStuff(ShipAPI ship, float amount, CombatEngineAPI engine) {
-        Color color;
-
-        // set up generators list
-        Map<ShipAPI, ShipAPI> storedGenerators; // nongenerator part -> shield generator
+    private ShipAPI findGenerator(ShipAPI ship, CombatEngineAPI engine) {
+        // First, try getting generator directly from the cached data
+        Map<FleetMemberAPI, ShipAPI> storedGenerators;
         if (engine.getCustomData().get(STORED_GENERATORS_KEY) instanceof Map) {
-            storedGenerators = (Map<ShipAPI, ShipAPI>) engine.getCustomData().get(STORED_GENERATORS_KEY);
+            storedGenerators = (Map<FleetMemberAPI, ShipAPI>) engine.getCustomData().get(STORED_GENERATORS_KEY);
         } else {
             storedGenerators = new HashMap<>();
         }
-
+        utilLog("storedGenerators.size(): "+storedGenerators.size());
         // find the generator
-        ShipAPI generator = storedGenerators.get(ship);
+        FleetMemberAPI shipsFleetMember = getFleetMember(ship);
+        ShipAPI generator = storedGenerators.get(shipsFleetMember);
         if (generator == null) {
-            for (ShipAPI check : CombatUtils.getShipsWithinRange(ship.getLocation(), 1000)) {
+            // Now try getting generator directly since we have no cached data
+            List<ShipAPI> shipsWithin1000range = CombatUtils.getShipsWithinRange(ship.getLocation(), 1000);
+
+            for (ShipAPI check : shipsWithin1000range) {
                 if (check.getHullSpec() != null
                         && check.getHullSpec().getHullId() != null
                         && check.getHullSpec().getHullId().equals(SHIELD_GENERATOR_ID)
-                        && (ship.equals(check.getParentStation())
-                        || (ship.getParentStation() != null
-                        && ship.getParentStation().equals(check.getParentStation())))) {
+                        && ship.getParentStation() != null
+                        && ship.getParentStation().equals(check.getParentStation())
+                ) {
```

Comment:
well, identical "extract these things" comment here as well

Reply:
done. actually somewhat simpler to skim through, but has added cognitive load of looking the values up...

Reply:
resolving